crates/net/src/rlpx/ecies.rs

```diff
@@ -0,0 +1,386 @@
+use aes::cipher::{KeyIvInit, StreamCipher};
+use bytes::BufMut;
+use ethereum_rust_core::{
+    rlp::{
+        decode::RLPDecode,
+        encode::RLPEncode,
+        error::RLPDecodeError,
+        structs::{Decoder, Encoder},
+    },
+    Signature, H128, H512, H520,
+};
+use k256::{
+    ecdsa::SigningKey,
+    elliptic_curve::sec1::{FromEncodedPoint, ToEncodedPoint},
+    EncodedPoint, PublicKey, SecretKey,
+};
+use keccak_hash::H256;
+use rand::{thread_rng, Rng};
+
+type Aes128Ctr64BE = ctr::Ctr64BE<aes::Aes128>;
+
+#[derive(Debug, Clone, Copy)]
+#[allow(unused)]
+pub(crate) struct HandshakeData {
+    pub remote_nonce: H256,
+    pub aes_key: H256,
+    pub mac_key: H256,
+}
+
+#[derive(Debug)]
+// TODO: refactor into two parts, one for the handshake and another one after that.
+pub(crate) struct RLPxConnection {
+    pub nonce: H256,
+    pub ephemeral_key: SecretKey,
+    pub handshake_data: Option<HandshakeData>,
+}
+
+impl RLPxConnection {
+    pub fn new(nonce: H256, ephemeral_key: SecretKey) -> Self {
+        Self {
+            nonce,
+            ephemeral_key,
+            handshake_data: None,
+        }
+    }
+
+    pub fn random() -> Self {
+        let mut rng = thread_rng();
+        Self::new(H256::random_using(&mut rng), SecretKey::random(&mut rng))
+    }
+
+    pub fn is_connected(&self) -> bool {
+        self.handshake_data.is_some()
+    }
+
+    pub fn encode_auth_message(
+        &self,
+        static_key: &SecretKey,
+        remote_static_pubkey: &PublicKey,
+        buf: &mut dyn BufMut,
+    ) {
+        let mut rng = rand::thread_rng();
+        let node_id = pubkey2id(&static_key.public_key());
+
+        let message_secret_key = SecretKey::random(&mut rng);
+        let message_secret = ecdh_xchng(&message_secret_key, remote_static_pubkey);
+
+        let signature = self.sign_shared_secret(message_secret.into());
+
+        let auth = AuthMessage::new(signature, node_id, self.nonce);
+        let padding_length = rng.gen_range(100..=300);
+
+        let mut encoded_auth_msg = vec![];
+        auth.encode(&mut encoded_auth_msg);
+        encoded_auth_msg.resize(encoded_auth_msg.len() + padding_length, 0);
+
+        let ecies_data_size = 65 + 16 + 32;
```

Comment:
shouldn't we comment the reason behind these numbers?

Reply:
Done!